src/core/reference/include/ngraph/runtime/reference/unique.hpp

```diff
@@ -0,0 +1,325 @@
+    /// duplicates found in the input tensor.
+    Count_t count = 1;
+    /// Indicates if this object points to a single value in the input tensor (rather than a slice of the tensor)
+    DescriptorType descriptor_type;
+};
+
+template <typename Index_t, typename Count_t>
+struct UniqueElements {
+    /// Contains descriptors of all elements in the input tensor. Possibly sorted by value.
+    std::vector<TensorSlice<Index_t, Count_t>> all_tensor_elements;
+    /// Subset of all tensor elements. First occurrences of the unique values.
+    std::vector<TensorSlice<Index_t, Count_t>> unique_tensor_elements;
+    /// Axis (optional). Used to gather unique elements over a given dimension.
+    int64_t axis = 0;
+};
+
+namespace {
+
+// Generates descriptors of slices or individual elems of the input tensor. This function returns a vector of
+// helper objects representing elements that the "unique" algorithm is supposed to process later.
+template <typename Index_t, typename Count_t>
+std::vector<TensorSlice<Index_t, Count_t>> generate_descriptors(const size_t count, const DescriptorType type) {
+    std::vector<TensorSlice<Index_t, Count_t>> descriptors;
+    descriptors.reserve(count);
+
+    for (Index_t i = 0; i < count; ++i) {
+        descriptors.emplace_back(i, type);
+    }
+
+    return descriptors;
+}
+
+// Returns indices of the first element of each tensor slice. The index is equal to a coordinate index.
+template <typename Index_t, typename Count_t>
+inline std::pair<size_t, size_t> first_elems_of_both_slices(const TensorSlice<Index_t, Count_t>& lhs,
+                                                            const TensorSlice<Index_t, Count_t>& rhs,
+                                                            const std::vector<size_t>& data_shape_strides,
+                                                            const int64_t axis) {
+    return {data_shape_strides[axis] * lhs.idx, data_shape_strides[axis] * rhs.idx};
+}
+
+template <typename Index_t, typename Count_t>
+inline size_t calc_slices_offset(const TensorSlice<Index_t, Count_t>& lhs,
+                                 const TensorSlice<Index_t, Count_t>& rhs,
+                                 const std::vector<size_t>& data_shape_strides,
+                                 const int64_t axis) {
+    const auto first_elem_indices = first_elems_of_both_slices(lhs, rhs, data_shape_strides, axis);
+    if (first_elem_indices.first > first_elem_indices.second) {
+        return first_elem_indices.first - first_elem_indices.second;
+    } else {
+        return first_elem_indices.second - first_elem_indices.first;
+    }
+}
+
+inline Shape slice_shape_to_iterate(Shape data_shape, const int64_t axis) {
+    data_shape.erase(data_shape.begin() + axis, data_shape.begin() + axis + 1);
+    return data_shape;
+}
+
+bool scalar_or_single_element(const Shape& s) {
+    return s.size() == 0 || std::all_of(std::begin(s), std::end(s), [](Shape::value_type d) {
```

Comment:
Partially applied. Using `ov::cmp::Equal` implies adding extra dependency between modules, therefore avoided.

Reply:
FYI: If required this file can be moved to core `dev_api` to make it more global for project. Right now is in inference lib because only there is used.